h2/utilities.py

```diff
@@ -319,13 +317,82 @@ def _check_host_authority_header(headers, hdr_validation_flags):
     # an :authority header nor a Host header.
     if not authority_present and not host_present:
         raise ProtocolError(
-            "Did not receive an :authority or Host header."
+            "Request header block must have an :authority or Host header."
         )
 
     # If we receive both headers, they should definitely match.
     if authority_present and host_present:
         if authority_header_val != host_header_val:
             raise ProtocolError(
-                "Received mismatched :authority and Host headers: %r / %r" %
-                (authority_header_val, host_header_val)
+                "Request header block must have matching :authority and "
+                "Host headers: %r / %r"
+                % (authority_header_val, host_header_val)
             )
+
+
+def _check_host_authority_header(headers, hdr_validation_flags):
+    """
+    Raises a ProtocolError if a header block arrives that does not contain an
+    :authority or a Host header, or if a header block contains both fields,
+    but their values do not match.
+    """
+    # We only expect to see :authority and Host headers on request header
+    # blocks that aren't trailers, so skip this validation if we're on the
+    # server side or looking at trailer blocks.
+    if hdr_validation_flags.is_client or hdr_validation_flags.is_trailer:
+        return headers
+
+    return _validate_host_authority_header(headers)
```

Comment:
Shiny, nice change! I had some vague concerns about what we were going to do in the absence of `yield from`, but changing this function to always return a generator without having any `yield` statements in it is a great solution to that problem. Well done!